File: samples/hos/elevator/comparison/elevator_naive.ml
(* elevator_naive.ml

   A deliberately flawed two-elevator controller in OCaml.

   To compile (requires ocaml + dune or direct build):
     ocamlfind ocamlopt -package threads.posix -linkpkg -thread \
       elevator_naive.ml -o elevator_naive

   Note: OCaml's module system and strong typing actually make some of the
   flaws from the Python version HARDER to write. I've annotated where
   OCaml's defaults push you away from the bugs, and where I had to go
   out of my way to reproduce them for parity with the Python version.
*)

(* ---------- Types ---------- *)

type direction = Up | Down | Idle

type door_state = Closed | Opening_door | Open_door | Closing_door

type motor_state = Stopped | Moving

type request = {
  floor: int;
  direction: direction;
  timestamp: float;
}

(* ---------- Door ---------- *)
(* FLAW 1: Door's state is mutable and publicly accessible.
   In idiomatic OCaml we'd hide this with a module signature. I've
   deliberately left it open to parallel the Python version. *)

module Door = struct
  type t = {
    car_id: int;
    mutable state: door_state;
    mutex: Mutex.t;
  }

  let make car_id = {
    car_id;
    state = Closed;
    mutex = Mutex.create ();
  }

  let open_ d =
    Mutex.lock d.mutex;
    if d.state = Closed then begin
      d.state <- Opening_door;
      (* FLAW 1: no check that motor is stopped *)
      Thread.delay 0.5;
      d.state <- Open_door
    end;
    Mutex.unlock d.mutex

  let close d =
    Mutex.lock d.mutex;
    if d.state = Open_door then begin
      d.state <- Closing_door;
      Thread.delay 0.5;
      d.state <- Closed
    end;
    Mutex.unlock d.mutex
end

(* ---------- Motor ---------- *)

module Motor = struct
  type t = {
    car_id: int;
    mutable state: motor_state;
    mutable current_floor: int;
    mutex: Mutex.t;
  }

  let make car_id = {
    car_id;
    state = Stopped;
    current_floor = 1;
    mutex = Mutex.create ();
  }

  let move_to m floor =
    Mutex.lock m.mutex;
    (* FLAW 2: no check that door is closed *)
    m.state <- Moving;
    let dir = if floor > m.current_floor then 1 else -1 in
    while m.current_floor <> floor do
      Thread.delay 0.3;
      m.current_floor <- m.current_floor + dir
    done;
    m.state <- Stopped;
    Mutex.unlock m.mutex
end

(* ---------- ElevatorCar ---------- *)

module Car = struct
  type t = {
    car_id: int;
    door: Door.t;
    motor: Motor.t;
    mutable queue: request list;
    mutex: Mutex.t;
    mutable running: bool;
    mutable emergency: bool;  (* FLAW 3: plain mutable flag, read without lock *)
  }

  let make car_id = {
    car_id;
    door = Door.make car_id;
    motor = Motor.make car_id;
    queue = [];
    mutex = Mutex.create ();
    running = true;
    emergency = false;
  }

  let add_request c req =
    Mutex.lock c.mutex;
    c.queue <- c.queue @ [req];
    Mutex.unlock c.mutex

  let rec run c =
    if not c.running then ()
    else begin
      (* FLAW 3: read emergency without lock, act on stale value *)
      if c.emergency then begin
        Door.open_ c.door;
        Thread.delay 0.1;
        run c
      end else begin
        let req = ref None in
        Mutex.lock c.mutex;
        (match c.queue with
         | [] -> ()
         | r :: rest -> req := Some r; c.queue <- rest);
        Mutex.unlock c.mutex;
        (match !req with
         | None -> Thread.delay 0.1
         | Some r ->
           (* FLAW 4: sequence enforced only by statement order *)
           Door.close c.door;
           Motor.move_to c.motor r.floor;
           Door.open_ c.door);
        run c
      end
    end

  (* FLAW 5: called by EmergencySystem directly, bypassing Dispatcher *)
  let emergency_stop c = c.emergency <- true

  let stop c = c.running <- false
end

(* ---------- Dispatcher ---------- *)

module Dispatcher = struct
  type t = {
    cars: Car.t array;
    mutable pending: request list;
    mutex: Mutex.t;
    mutable running: bool;
  }

  let make cars = {
    cars;
    pending = [];
    mutex = Mutex.create ();
    running = true;
  }

  let request d floor direction =
    Mutex.lock d.mutex;
    d.pending <- d.pending @ [{floor; direction; timestamp = Unix.gettimeofday ()}];
    Mutex.unlock d.mutex

  let rec run d =
    if not d.running then ()
    else begin
      let req = ref None in
      Mutex.lock d.mutex;
      (match d.pending with
       | [] -> ()
       | r :: rest -> req := Some r; d.pending <- rest);
      Mutex.unlock d.mutex;
      (match !req with
       | None -> Thread.delay 0.1
       | Some r ->
         (* FLAW 6: stale snapshot of car positions *)
         let best = ref d.cars.(0) in
         let best_dist = ref (abs (d.cars.(0).motor.current_floor - r.floor)) in
         for i = 1 to Array.length d.cars - 1 do
           let dist = abs (d.cars.(i).motor.current_floor - r.floor) in
           if dist < !best_dist then begin
             best := d.cars.(i);
             best_dist := dist
           end
         done;
         Car.add_request !best r);
      run d
    end
end

(* ---------- EmergencySystem ---------- *)

module Emergency = struct
  type t = {
    cars: Car.t array;
    dispatcher: Dispatcher.t;  (* held but unused *)
  }

  let make cars dispatcher = { cars; dispatcher }

  (* FLAW 5: reaches into cars directly, bypassing dispatcher *)
  let trigger e = Array.iter Car.emergency_stop e.cars
end

(* ---------- Build and run ---------- *)

let build_system () =
  let cars = [| Car.make 1; Car.make 2 |] in
  let dispatcher = Dispatcher.make cars in
  let emergency = Emergency.make cars dispatcher in
  let _t1 = Thread.create Car.run cars.(0) in
  let _t2 = Thread.create Car.run cars.(1) in
  let _td = Thread.create Dispatcher.run dispatcher in
  (cars, dispatcher, emergency)

let () =
  let (cars, dispatcher, _emergency) = build_system () in
  Dispatcher.request dispatcher 5 Up;
  Thread.delay 0.1;
  Dispatcher.request dispatcher 3 Down;
  Thread.delay 5.0;
  Printf.printf "Car 1 floor: %d\n" cars.(0).motor.current_floor;
  Printf.printf "Car 2 floor: %d\n" cars.(1).motor.current_floor

File: samples/hos/elevator/comparison/elevator_naive.rs
// elevator_naive.rs
//
// A deliberately flawed two-elevator controller in Rust.
//
// IMPORTANT: Writing this was genuinely hard. The borrow checker refuses
// several of the most natural "bad" patterns from the Python version.
// Specifically, the sibling back-channel in Bug 5 (EmergencySystem holding
// Vec<Arc<ElevatorCar>>) requires Arc<Mutex<...>> everywhere, because
// without it you cannot share mutable references across threads at all.
// So even the "broken" version pays a correctness tax that Python doesn't.
//
// This tells you something: Rust makes it HARDER to write the broken version
// and arguably impossible to write the Python-style broken version verbatim.
// Many of the flaws are attenuated even in the naive version.
//
// Build: cargo new elevator_naive && cargo add --features none
// Then put this as src/main.rs and run: cargo run

use std::collections::VecDeque;
use std::sync::{Arc, Mutex};
use std::thread;
use std::time::{Duration, SystemTime};

#[derive(Debug, Clone, Copy, PartialEq)]
enum Direction { Up, Down, Idle }

#[derive(Debug, Clone, Copy, PartialEq)]
enum DoorState { Closed, Opening, Open, Closing }

#[derive(Debug, Clone, Copy, PartialEq)]
enum MotorState { Stopped, Moving }

#[derive(Debug, Clone)]
struct Request {
    floor: i32,
    #[allow(dead_code)]
    direction: Direction,
    #[allow(dead_code)]
    timestamp: SystemTime,
}

struct Door {
    #[allow(dead_code)]
    car_id: u32,
    state: DoorState,
}

impl Door {
    fn new(car_id: u32) -> Self { Self { car_id, state: DoorState::Closed } }

    fn open(&mut self) {
        if self.state == DoorState::Closed {
            self.state = DoorState::Opening;
            // FLAW 1: no check that motor is stopped
            thread::sleep(Duration::from_millis(500));
            self.state = DoorState::Open;
        }
    }

    fn close(&mut self) {
        if self.state == DoorState::Open {
            self.state = DoorState::Closing;
            thread::sleep(Duration::from_millis(500));
            self.state = DoorState::Closed;
        }
    }
}

struct Motor {
    #[allow(dead_code)]
    car_id: u32,
    state: MotorState,
    current_floor: i32,
}

impl Motor {
    fn new(car_id: u32) -> Self {
        Self { car_id, state: MotorState::Stopped, current_floor: 1 }
    }

    fn move_to(&mut self, floor: i32) {
        // FLAW 2: no check that door is closed
        self.state = MotorState::Moving;
        let dir = if floor > self.current_floor { 1 } else { -1 };
        while self.current_floor != floor {
            thread::sleep(Duration::from_millis(300));
            self.current_floor += dir;
        }
        self.state = MotorState::Stopped;
    }
}

struct ElevatorCar {
    car_id: u32,
    door: Mutex<Door>,
    motor: Mutex<Motor>,
    queue: Mutex<VecDeque<Request>>,
    running: Mutex<bool>,
    emergency: Mutex<bool>,  // FLAW 3: would be AtomicBool in real Rust
}

impl ElevatorCar {
    fn new(car_id: u32) -> Self {
        Self {
            car_id,
            door: Mutex::new(Door::new(car_id)),
            motor: Mutex::new(Motor::new(car_id)),
            queue: Mutex::new(VecDeque::new()),
            running: Mutex::new(true),
            emergency: Mutex::new(false),
        }
    }

    fn add_request(&self, req: Request) {
        self.queue.lock().unwrap().push_back(req);
    }

    fn current_floor(&self) -> i32 {
        self.motor.lock().unwrap().current_floor
    }

    fn run(self: Arc<Self>) {
        loop {
            if !*self.running.lock().unwrap() { break; }

            // FLAW 3: TOCTOU - read emergency, release lock, then act
            let is_emergency = *self.emergency.lock().unwrap();
            if is_emergency {
                self.door.lock().unwrap().open();
                thread::sleep(Duration::from_millis(100));
                continue;
            }

            let req = self.queue.lock().unwrap().pop_front();
            match req {
                None => thread::sleep(Duration::from_millis(100)),
                Some(r) => {
                    // FLAW 4: sequence by statement order, not structure.
                    // The borrow checker ensures the locks are released
                    // between calls, but it does NOT enforce the ordering
                    // of operations across lock acquisitions.
                    self.door.lock().unwrap().close();
                    self.motor.lock().unwrap().move_to(r.floor);
                    self.door.lock().unwrap().open();
                }
            }
        }
    }

    // FLAW 5: called by EmergencySystem directly. Still possible in Rust
    // because we gave out Arc<ElevatorCar>.
    fn emergency_stop(&self) {
        *self.emergency.lock().unwrap() = true;
    }

    #[allow(dead_code)]
    fn stop(&self) {
        *self.running.lock().unwrap() = false;
    }
}

struct Dispatcher {
    cars: Vec<Arc<ElevatorCar>>,
    pending: Mutex<VecDeque<Request>>,
    running: Mutex<bool>,
}

impl Dispatcher {
    fn new(cars: Vec<Arc<ElevatorCar>>) -> Self {
        Self {
            cars,
            pending: Mutex::new(VecDeque::new()),
            running: Mutex::new(true),
        }
    }

    fn request(&self, floor: i32, direction: Direction) {
        self.pending.lock().unwrap().push_back(Request {
            floor,
            direction,
            timestamp: SystemTime::now(),
        });
    }

    fn run(self: Arc<Self>) {
        loop {
            if !*self.running.lock().unwrap() { break; }

            let req = self.pending.lock().unwrap().pop_front();
            match req {
                None => thread::sleep(Duration::from_millis(100)),
                Some(r) => {
                    // FLAW 6: stale snapshot of floors
                    let best = self.cars.iter()
                        .min_by_key(|c| (c.current_floor() - r.floor).abs())
                        .unwrap();
                    best.add_request(r);
                }
            }
        }
    }
}

struct EmergencySystem {
    cars: Vec<Arc<ElevatorCar>>,
    #[allow(dead_code)]
    dispatcher: Arc<Dispatcher>,  // held but unused
}

impl EmergencySystem {
    fn new(cars: Vec<Arc<ElevatorCar>>, dispatcher: Arc<Dispatcher>) -> Self {
        Self { cars, dispatcher }
    }

    // FLAW 5: reaches into cars directly
    fn trigger(&self) {
        for c in &self.cars {
            c.emergency_stop();
        }
    }
}

fn main() {
    let cars: Vec<Arc<ElevatorCar>> = (1..=2)
        .map(|i| Arc::new(ElevatorCar::new(i)))
        .collect();

    let dispatcher = Arc::new(Dispatcher::new(cars.clone()));
    let _emergency = EmergencySystem::new(cars.clone(), dispatcher.clone());

    for c in &cars {
        let c2 = c.clone();
        thread::spawn(move || c2.run());
    }
    let d2 = dispatcher.clone();
    thread::spawn(move || d2.run());

    dispatcher.request(5, Direction::Up);
    thread::sleep(Duration::from_millis(100));
    dispatcher.request(3, Direction::Down);
    thread::sleep(Duration::from_secs(5));

    for c in &cars {
        println!("Car {} floor: {}", c.car_id, c.current_floor());
    }
}
